### DISCUSSION TITLE
How to mix command criteria `[class="XyZ"]` and `exec`?

### DISCUSSION THREAD

**Discussion by a-pav** [i3/i3#5786] [Q&A] [answered]
I'm trying to do this
```
bindsym ctrl+shift+f [class="firefox-aurora"] exec --no-startup-id "xdotool key --clearmodifiers ctrl+shift+f Down"
```
and it doesn't work. While doing this
```
bindsym ctrl+shift+f [class="firefox-aurora"] border normal
```
works.

Since in [the related documentation](https://i3wm.org/docs/userguide.html#command_criteria) also there's no example including `exec`, I'm wondering if that's  generally im/possible?

**Comment by orestisfl** [comment]
~~Yes, exec commands do not consider command criteria, I am not sure what would that mean, what does it mean to execute a command "for" a window?~~

In your case, if I understand the intention correctly, try:
```
bindsym --release ctrl+shift+f [class="firefox-aurora"] focus, exec --no-startup-id "xdotool key --clearmodifiers ctrl+shift+f Down"
```

**Comment by a-pav** [comment]
> what does it mean to execute a command "for" a window?

It's _the basic_ for defining contextual keybindings. Contextual keybindings act only within their context (that is, when their associated window is focused, for example.) Otherwise, triggering them is no-op. This should make it possible to define the exact same keybinding for different contexts. Think of what pressing `alt+f` does in firefox, and what it does in a file manager; They are the same keybinding, but do different things. Yet, from i3 point-of-view, this is currently considered duplicate definition: `ERROR: Duplicate keybinding in config file`. (To my surprise, using i3 way of command criteria, the `[class="XyZ"]`, doesn't help this case.)

This is certainly not a new way of thinking about keybindings. For example:
  - VSCode uses the [`"when":` clause](https://code.visualstudio.com/api/references/when-clause-contexts) in their keybindings definition.
  - AutoHotKey, a Windows scripting language, used to have [`#IfWin<criteria>` clases](https://www.autohotkey.com/docs/v1/lib/_IfWinActive.htm), (now [`#HotIf`](https://www.autohotkey.com/docs/v2/lib/_HotIf.htm).)

**Reply by orestisfl** [reply, marked as answer]
Right, the i3-specific way to do that is:

```
bindsym $mod+g [con_id=__focused__ class="firefox"] exec notify-send "this is a browser", [con_id=__focused__ class="Gnome-terminal"] exec notify-send "this is a terminal"
```

This will execute only the first command on firefox and only the second on gnome terminal.

**Reply by a-pav** [reply]
Then `[command-criteria]` and `exec` do in fact work together, or at least with something like `con_id=__focused__` they do.
That's basically what I wanted to do. Thanks for the help.

Although, making things work with `xdotool` (rather than `notify-send`) seem to remain a challenge.

**Reply by orestisfl** [reply]
> Then [command-criteria] and exec do in fact work togethe

Yes, I was mistaken before, but after finding https://github.com/i3/i3/issues/3903 again, I remembered it's supposed to work.

For `xdotool`, try `bindsym --release`.

**Reply by a-pav** [reply]
BTW, I forgot to mention, when I trigger the keybinding snippet you provided in your answer on an empty workspace, both command are executed! Obviously no window is focused in an empty workspace, so `con_id=__focused__` shouldn't match any window at all. Maybe I should report this on #3903? I'm using `i3 version 4.20.1 (2021-11-03)`

> For `xdotool`, try `bindsym --release`

Yeah, that, or the many tricks in the book of `xdotool`.

Since I'm bumping this up again, please let me ask another question. Is there an i3-specific way to check for the existence of a specific window, with the criteria like `class`, `name`, etc. ? I already have some idea how to do that with `xdotool search` or `xprope -name` I guess, or even `i3-save-tree`

**Reply by a-pav** [reply]
> BTW, I forgot to mention, when I trigger the keybinding snippet you provided in your answer on an empty workspace, both command are executed! Obviously no window is focused in an empty workspace, so `con_id=__focused__` shouldn't match any window at all. Maybe I should report this on #3903? I'm using `i3 version 4.20.1 (2021-11-03)`

I just realized in addition to the issue I previously pointed out, when you set a keybinding like this, it becomes unavailable everywhere else, i.e. even the windows which support that keybinding natively, become unable to capture it.

For example, by default, pressing `alt+m` opens the _Media menu in VLC, So if I set a keybinding like this:
```sh
bindsym $alt+m [con_id=__focused__ class="firefox-aurora"] exec notify-send "this is a browser", [con_id=__focused__ class="Gnome-terminal"] exec notify-send "this is a terminal"
```
pressing `$alt+m` no longer does its default job on VLC, and **this is an undesired behavior**.

**Reply by orestisfl** [reply]
> BTW, I forgot to mention, when I trigger the keybinding snippet you provided in your answer on an empty workspace, both command are executed! Obviously no window is focused in an empty workspace, so con_id=__focused__ shouldn't match any window at all. Maybe I should report this on https://github.com/i3/i3/issues/3903? I'm using i3 version 4.20.1 (2021-11-03)

Interesting, this looks like a bug, thanks for noticing, I opened https://github.com/i3/i3/issues/6037 and a fix is on the way.

> Since I'm bumping this up again, please let me ask another question. Is there an i3-specific way to check for the existence of a specific window, with the criteria like class, name, etc. ? I already have some idea how to do that with xdotool search or xprope -name I guess, or even i3-save-tree

You would have to use the [IPC interface](https://i3wm.org/docs/ipc.html) for that.

> I just realized in addition to the issue I previously pointed out, when you set a keybinding like this, it becomes unavailable everywhere else, i.e. even the windows which support that keybinding natively, become unable to capture it.

Conditional bindsyms are intentionally left out, see discussion in https://github.com/i3/i3/issues/4768. Ideally, we would want a generic key binding daemon to implement more complex binding logic. I think [autokey](https://github.com/autokey/autokey) can do that but I haven't investigated deeper. If you find something, please do let me know.

There could be some value in bindings that do not consume the event, that came up in https://github.com/i3/i3/discussions/5974#discussioncomment-9260264 recently

**Reply by a-pav** [reply]
> Interesting, this looks like a bug, thanks for noticing, I opened https://github.com/i3/i3/issues/6037 and a fix is on the way.

Great.

> You would have to use the [IPC interface](https://i3wm.org/docs/ipc.html) for that.

I'm using the output of `wmctrl -lx`. It's fine for simple window matching based on class and title.

> Conditional bindsyms are intentionally left out, see discussion in https://github.com/i3/i3/issues/4768. Ideally, we would want a generic key binding daemon to implement more complex binding logic. I think [autokey](https://github.com/autokey/autokey) can do that but I haven't investigated deeper. If you find something, please do let me know.

I don't understand why i3 team (often) dismisses keybinding related issues and requests by mentioning other, **external** solutions. The **built-in** solution that i3 currently offers is robust enough, far better than anything else I've seen on Linux. I haven't tried `Autokey` but tried `sxhkd` awhile back. There's nothing in `sxhkd` that a user can't do in i3 more reliably. The only interesting feature of `sxhkd` was its _chords_, but I'm producing the same effect in i3 by moving through  `mode`s. The configuration for that in i3 looks a bit more complicated than it would have in sxhkd, but I'm fine with it. Once I have it working, making changes is easy.

> There could be some value in bindings that do not consume the event, that came up in https://github.com/i3/i3/discussions/5974#discussioncomment-9260264 recently

I can't imagine what value if conditional keybindings are not available. (I don't share the requirement of the OP in that discussion. I never had a problem with hover-to-focus mechanism of i3.) Besides, I wonder if that would make it impossible to _override_ a default keybinding for an app, since both the default and user defined behaviors are going to be triggered.

**Reply by a-pav** [reply]
I think it's more worthwhile to focus on conditional keybindings. Consider the following please.

If I have a keybinding like this:
```sh
# for browser
bindsym $alt+m [con_id=__focused__ class="firefox-aurora"] exec command_1
```
I still should be able to define `bindsym $alt+m` for another condition(s) or no condition at all, like:
```sh
# for terminal
bindsym $alt+m [con_id=__focused__ class="Gnome-terminal"] exec command_2
# for global/system
bindsym $alt+m exec command_3
```
All of these three `bindsym $alt+m` definitions (and more) should be able to coexist inside an i3 configuration file at once.

If this was the design, there would have been no need to create https://github.com/i3/i3/issues/6037, and also a lot easier to think of a solution for OP in https://github.com/i3/i3/discussions/5974#discussioncomment-9260264.
